### ISSUE TITLE
Sampling output image visualization problem?

### ISSUE THREAD

**Issue by junxiant** [WuJunde/MedSegDiff#26]
Hello,

I'm currently training this model on my own dataset, I have created a separate dataloader python file, the file and folder structure of the dataset is the same as ISIC. No other code other than segmentation_train.py and segmentation_sample.py was changed just to load the data. The model is trained for 30000 steps so far. But when I tried to use the segmentation_sample.py for the test images, I am getting these masks. 

Are these mask outputs normal for this model?
![89_16_output_ens](https://user-images.githubusercontent.com/25401067/216563989-11eba24e-3923-45ef-975f-b3a696067ffc.jpg)
![89_24_output_ens](https://user-images.githubusercontent.com/25401067/216564445-515f57d0-b911-40ad-8cc9-97bb1277c027.jpg)

python 3.8.16
torch 1.13.1
torchvision 0.14.1
torchsummary 1.5.1
opencv 4.7.0.68
scikit-image 0.19.3

**Comment by WuJunde** [comment]
is this the ensemble result saved in your result file? Some others met the similar problem before https://github.com/WuJunde/MedSegDiff/issues/6.

**Comment by junxiant** [comment]
Yes, using the savedmodel030000.pt and these images are named ''{slice_id}_output_ens.jpg'

**Comment by junxiant** [comment]
This is one of the predicted masks after I trained the model on ISIC dataset

![image](https://user-images.githubusercontent.com/25401067/216875482-c93712ef-1f1a-49fc-880a-9b495c4c8456.png)

**Comment by WuJunde** [comment]
it's close. Note that MedSegDiff will outcome uncertianty-aware segmentation, because of the uncertianty nature of medical image segmentation(check: Learning calibrated medical image segmentation via multi-rater agreement modeling(CVPR 2021), Learning self-calibrated optic disc and cup segmentation from multi-rater annotations(MICCAI 2022)). but your prediction seems a little too much, you may want to generate more confident results by more training or ensemble.

if you want to obtain the binary results, you can threshold and postprocess this prediction. I used the postprocess config in nnUNet for example.

**Comment by WangPing521** [comment]
> it's close. Note that MedSegDiff will outcome uncertianty-aware segmentation, because of the uncertianty nature of medical image segmentation(check: Learning calibrated medical image segmentation via multi-rater agreement modeling(CVPR 2021), Learning self-calibrated optic disc and cup segmentation from multi-rater annotations(MICCAI 2022)). but your prediction seems a little too much, you may want to generate more confident results by more training or ensemble.
> 
> if you want to obtain the binary results, you can threshold and postprocess this prediction. I used the postprocess config in nnUNet for example.

I want to know what is the input of diffusion model, in order to get a segmentation? For example, we have a 5-class within an image. Then we have two choice, one is to use the one-hot (1, 5, h, w) where the values are 0 and 1, suppose batch_size is 1, as the input of diffusion model; another is to use the mask(1,1,h,w) where the values are 0, 1, 2, 3, 4, as the input of diffusion model. Which one is you choice? In this diffusion on segmentation rather than image case, what are the output during the reverse stage? How to map the sampled output to the labels(0 and1, or 0,1,2,3,4)?

If i am wrong above, as you said "MedSegDiff will outcome uncertianty-aware segmentation", the output is probability map like the output of normal segmentation network? Can you give me some hints about this? I am very confusing about the diffusion on segmentation rather than image. 

I would very appreciate if you can help.

**Comment by WuJunde** [comment]
@WangPing521 It depends on your dataset, both of them can work. if it was me, i will use (1, 5, h, w). diffusion model outputs a residual noise. Then substract it from the last noisy result(segmentation) to get a clearer one. The iteration enables it to generate a segmentation from a gaussian noise conditioned by the raw image. you can refer to our paper for more details.

**Comment by WangPing521** [comment]
> 

Thanks for your reply!

I have  a naive question for your input of code, "self.run_step(batch, cond) ", here, Does the batch mean the label, the cond mean the original image? Can you show the dimensionality of these two tensors in a (N, C, H, W) way?

**Comment by hding9** [comment]
@WangPing521 If I understand correctly, the first step in ["run_step"](https://github.com/WuJunde/MedSegDiff/blob/da876ed7a0447ae860edc68d2643f9dc99109c9a/guided_diffusion/train_util.py#L203) concatenates the image and mask together, and leave "cond" as an empty dictionary. For ISIC dataset, I think the image size is [B, 3, H, W] and the mask is [B, 1, H, W].
